### PR TITLE
Upgrade IceUtil::Timer to use std::chrono::duration

### PR DESCRIPTION
**Pull request by pepone**
This PR upgrades IceUtil::Timer to not depend on IceUtil::Time, which is replaced by std::chrono::duration and std::chrono::steady_clock::time_point.

I also updated the timer to use std::thread

### CODE REVIEW

File: cpp/include/IceUtil/Timer.h [hunk 1]
```diff
@@ -33,41 +29,78 @@ class ICE_API TimerTask
...
+    {
+        std::lock_guard lock(_mutex);
+        if(_destroyed)
+        {
+            throw IllegalArgumentException(__FILE__, __LINE__, "timer destroyed");
+        }
+
+        auto now = std::chrono::steady_clock::now();
+        auto time = now + delay;
+        if(delay < std::chrono::nanoseconds::zero())
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/include/IceUtil/Timer.h:88]: Yes

File: cpp/include/IceUtil/Timer.h [hunk 2]
```diff
@@ -79,63 +112,39 @@ class ICE_API Timer : public Thread
...
-            return lhs.get() < rhs.get();
+            if (scheduledTime < other.scheduledTime)
+            {
+                return true;
+            }
+            else if (scheduledTime > other.scheduledTime)
+            {
+                return false;
+            }
+            return task < other.task;
```
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/include/IceUtil/Timer.h:135]: Not sure, the suggestion is a bit more compact but the original code does a comparison less when `scheduledTime < other.scheduledTime;`

File: cpp/include/IceUtil/Timer.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/IceUtil/Timer.h:88]: It should be <= 0? 0 doesn't seem correct either.

File: cpp/include/IceUtil/Timer.h [hunk 3]
```diff
@@ -33,41 +30,78 @@ class ICE_API TimerTask
...
+        }
 
-    //
-    // Schedule a task for repeated execution with the given delay
-    // between each execution.
-    //
-    void scheduleRepeated(const TimerTaskPtr&, const IceUtil::Time&);
+        auto now = std::chrono::steady_clock::now();
+        auto time = now + delay;
+        if (delay < std::chrono::nanoseconds::zero())
```
Comment (bernardnormier) [inline cpp/include/IceUtil/Timer.h:58]: It's odd to perform this check after computing now and time.

File: cpp/include/IceUtil/Timer.h [hunk 4]
```diff
@@ -33,41 +30,78 @@ class ICE_API TimerTask
...
+        {
+            throw IllegalArgumentException(__FILE__, __LINE__, "timer destroyed");
+        }
 
-    //
-    // Schedule a task for repeated execution with the given delay
-    // between each execution.
-    //
-    void scheduleRepeated(const TimerTaskPtr&, const IceUtil::Time&);
+        auto now = std::chrono::steady_clock::now();
```
Comment (bernardnormier) [inline cpp/include/IceUtil/Timer.h:56]: We only use `now` once on the line below.

File: cpp/src/Ice/Instance.cpp [hunk 5]
```diff
@@ -185,30 +185,16 @@ class Timer : public IceUtil::Timer
...
-    static TimerPtr create()
-    {
-        auto timer = shared_ptr<Timer>(new Timer());
-        timer->start();
-        return timer;
-    }
-
-    static TimerPtr create(int priority)
+    Timer() :
+        _hasObserver(0)
```
Comment (bernardnormier) [inline cpp/src/Ice/Instance.cpp:189]: should be false, not 0.

**Review by bernardnormier — APPROVED** [review]
I assume you plan to remove IceUtil::Time in a follow-up PR?

File: cpp/include/IceUtil/Timer.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/include/IceUtil/Timer.h:88]: We use 0 in a few places to schedule something on the timer thread immediately:

```
src/IceStorm/NodeI.cpp
917:            _timer->schedule(_mergeContinueTask, chrono::seconds::zero());

src/IceLocatorDiscovery/PluginI.cpp
421:    if(_timeout < chrono::milliseconds::zero())
429:    if(_retryDelay < chrono::milliseconds::zero())
431:        _retryDelay = chrono::milliseconds::zero();
```

And we have tests for using 0
```
test/IceUtil/timer/Client.cpp
172:            timer->schedule(task, chrono::seconds::zero());
181:                timer->schedule(task, chrono::seconds::zero());
255:            timer->schedule(destroyTask, chrono::seconds::zero());
259:                timer->schedule(destroyTask, chrono::seconds::zero());
```

**Comment by externl** [comment]
> > Do we also need to update the props to remove the unused configuration?
> 
> What props?

Nvm, I was thinking threads.

**Comment by externl** [comment]
Looks like there are still come CI failures.

**Review by externl — APPROVED** [review]

File: cpp/include/IceUtil/Timer.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/include/IceUtil/Timer.h:86]: I misunderstood the purpose of the original code, it was about checking for overflows

See https://github.com/zeroc-ice/ice/commit/8e4d4c244e2560ed0d9957a36f8eade79e455c0d

**Pull request merged**